astro/cli/astro/cmd/cmd.go

```diff
@@ -0,0 +1,291 @@
+/*
+ *  Copyright (c) 2018 Uber Technologies, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package cmd contains the source for the `astro` command line tool
+// that operators use to interact with the project.
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/uber/astro/astro"
+	"github.com/uber/astro/astro/conf"
+	"github.com/uber/astro/astro/logger"
+
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	// silence trace info from terraform/dag by default
+	log.SetOutput(ioutil.Discard)
+}
+
+// AstroCLI is the main CLI program, where flags and state are stored for the
+// running program.
+type AstroCLI struct {
+	stdin  io.Reader
+	stdout io.Writer
+	stderr io.Writer
+
+	project *astro.Project
+	config  *conf.Project
+
+	// these values are filled in based on runtime flags
+	flags struct {
+		detach            bool
+		moduleNamesString string
+		trace             bool
+		userCfgFile       string
+		verbose           bool
+
+		// projectFlags are special in that the actual flags are dynamic, based
+		// on the astro project configuration loaded.
+		projectFlags []*projectFlag
+	}
+
+	commands struct {
+		root  *cobra.Command
+		plan  *cobra.Command
+		apply *cobra.Command
+	}
+}
+
+// NewAstroCLI creates a new AstroCLI.
+func NewAstroCLI(opts ...Option) (*AstroCLI, error) {
+	cli := &AstroCLI{
+		stdin:  os.Stdin,
+		stdout: os.Stdout,
+		stderr: os.Stderr,
+	}
+
+	if err := cli.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	// Set up Cobra commands and structure
+	cli.createRootCommand()
```

Comment:
Chatted offline. We have to keep them as struct methods. Because all the subcommands have dependencies on `CLI` parent struct. For example, `plan  *cobra.Command` depends on `cli.runPlan`.

Reply:
> I always feel confused by the struct/file/directory names in `astro`.

These are really good comments. Let's file them away in an issue so we can fix this later. Maybe it could be a good task for you to take on? :-)